Improve message when a path contains illegal characters

papeh opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Home/issues/11764

Regression? Last working version:

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->

* When Path.Combine throws an ArgumentException, wrap it in an ArgumentException with a helpful message.

Addresses the stacks in https://github.com/NuGet/Home/issues/8373 and https://github.com/NuGet/Home/issues/11764; there may be others that these changes do not address.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

papeh committed:
Improve message when a path contains illegal characters

Addresses the stacks in #8373; there may be others that these changes do not address.

ghost commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/signed)](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=4586) <br/>All CLA requirements met.

smaudet commented:
@papeh Does this fix the error with an otherwise OK path? In my case at least my path was fine, I never went as far to hunt down the specific message but simply suppressing the error allowed the restore to proceed, indicating to me more a logic bug than simply a need to log a more verbose message.

papeh committed:
Add another unit test.

The other methods I altered were not unit tested before,
so no tests were added for them.

nkolev92 commented:
@papeh Why is Path.Combine throwing in your scenario? What's the problem with the path you are passing in?

papeh commented:
@nkolev92 here is my output before the change:
```
C:\repos\repo\build>"C:\repos\repo\build\/nuget.exe" install BuildTasks -verbosity detail -excludeVersion -version 2.3.1 -solutionDirectory "C:\repos\repo\build\"
NuGet Version: 6.1.0.106
Illegal characters in path.
System.ArgumentException: Illegal characters in path.
   at System.IO.Path.CheckInvalidPathChars(String path, Boolean checkAdditional)
   at System.IO.Path.Combine(String path1, String path2)
   at NuGet.CommandLine.InstallCommand.CalculateEffectiveSettings()
   at NuGet.CommandLine.InstallCommand.ExecuteCommandAsync()
   at NuGet.CommandLine.Command.Execute()
   at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args)
```
The problem in my case is that the solution is actually at `C:\repos\repo\src\Solution.sln`. The specified solution directory, `C:\repos\repo\build\` is a legal path, so I assume that it's throwing because some other code has changed the path from the command line to something illegal, but I don't know what (the message doesn't say). The updated message will at least let users know which path has illegal characters.

nkolev92 commented:
Can you post your error the after the change too?

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs:
```diff
@@ -112,7 +112,15 @@ private void CalculateEffectiveSettings()
+                }
+                catch (ArgumentException e)
+                {
+                    throw new ArgumentException($"Invalid Solution Directory: '{SolutionDirectory}'", e);
```
Please use a resource string for user facing strings.

Also, please don't capitalize S and D and instead do `Invalid solution directory`

papeh replied:
done

papeh committed:
Internationalize strings

papeh commented:
> Can you post your error the after the change too?
```
C:\repos\repo\build>"C:\repos\repo\build\/nuget.exe" install BuildTasks -verbosity detail -excludeVersion -version 2.3.1 -solutionDirectory "C:\repos\repo\build\"
NuGet Version: 6.3.0.32767
Invalid solution directory: 'C:\repos\repo\build"'
  Illegal characters in path.
System.ArgumentException: Invalid solution directory: 'C:\repos\repo\build"' ---> System.ArgumentException: Illegal characters in path.
   at System.IO.Path.CheckInvalidPathChars(String path, Boolean checkAdditional)
   at System.IO.Path.Combine(String path1, String path2)
   at NuGet.CommandLine.InstallCommand.CalculateEffectiveSettings()
   --- End of inner exception stack trace ---
   at NuGet.CommandLine.InstallCommand.CalculateEffectiveSettings()
   at NuGet.CommandLine.InstallCommand.ExecuteCommandAsync()
   at NuGet.CommandLine.Command.Execute()
   at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args)
```

Fixing the trailing `"` should be trivial; is that allowed to go in this same PR?

papeh committed:
Trim quotes before combining paths

* Trim quotes before combining paths passed as command line args
* Don't trim trailing directory separators; Path.Combine handles these
* Consolidate a pair of identical return statements

papeh commented:
I just realized that, rather than trimming `"` from paths before passing them to `Path.Combine`, it may be better to trim `"` from all paths as soon as they're passed on the command line, since quoting paths is common on Windows, and the command line processor is retaining them in some cases when it shouldn't. I'm happy to create a separate issue and PR for this, but only if the maintainers like the idea.

nkolev92 commented:
yeah, that makes sense. Wonder if it's the parser that has these issues? Why isn't the end `"` trimmed, but the starting one is? 

Can you try *not* passing solution directory as the final argument?

papeh commented:
Yup, the problem seems to be with the parser:
```
C:\repos\repo\build>"C:\repos\repo\build\/nuget.exe" install BuildTasks -verbosity detail -solutionDirectory "C:\repos\repo\build\" -excludeVersion -version 2.3.1
NuGet Version: 6.3.0.32767
Invalid solution directory: 'C:\repos\repo\build" -excludeVersion -version 2.3.1'
  Illegal characters in path.
System.ArgumentException: Invalid solution directory: 'C:\repos\repo\build" -excludeVersion -version 2.3.1' ---> System.ArgumentException: Illegal characters in path.
   at System.IO.Path.CheckInvalidPathChars(String path, Boolean checkAdditional)
   at System.IO.Path.Combine(String path1, String path2)
   at NuGet.CommandLine.InstallCommand.CalculateEffectiveSettings()
   --- End of inner exception stack trace ---
   at NuGet.CommandLine.InstallCommand.CalculateEffectiveSettings()
   at NuGet.CommandLine.InstallCommand.ExecuteCommandAsync()
   at NuGet.CommandLine.Command.Execute()
   at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args)
```

And with no `solutionDirectory`, just for fun:
```
C:\repos\repo\build>"C:\repos\repo\build\/nuget.exe" install BuildTasks -verbosity detail -excludeVersion -version 2.3.1
NuGet Version: 6.3.0.32767
Feeds used:
  C:\Users\papeh\.nuget\packages\
  https://api.nuget.org/v3/index.json
  C:\Dev\nugetpackages
  C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\



Attempting to gather dependency information for package 'BuildTasks.2.3.1' with respect to project 'C:\repos\repo\build', targeting 'Any,Version=v0.0'
Missing C:\Users\papeh\.nuget\packages\buildtasks\2.3.1\buildtasks.2.3.1.nupkg
Missing C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\buildtasks\2.3.1\buildtasks.2.3.1.nupkg
  GET https://api.nuget.org/v3/registration5-gz-semver2/buildtasks/index.json
  OK https://api.nuget.org/v3/registration5-gz-semver2/buildtasks/index.json 100ms
Total number of results gathered : 4
Package 'BuildTasks 2.3.1' is not found in the following primary source(s): 'C:\Users\papeh\.nuget\packages\,https://api.nuget.org/v3/index.json,C:\Dev\nugetpackages,C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\'. Please verify all your online package sources are available (OR) package id, version are specified correctly.
System.InvalidOperationException: Package 'BuildTasks 2.3.1' is not found in the following primary source(s): 'C:\Users\papeh\.nuget\packages\,https://api.nuget.org/v3/index.json,C:\Dev\nugetpackages,C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\'. Please verify all your online package sources are available (OR) package id, version are specified correctly.
   at NuGet.PackageManagement.ResolverGather.<GatherAsync>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at NuGet.PackageManagement.ResolverGather.<GatherAsync>d__21.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at NuGet.PackageManagement.NuGetPackageManager.<PreviewInstallPackageAsync>d__76.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at NuGet.PackageManagement.NuGetPackageManager.<PreviewInstallPackageAsync>d__75.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at NuGet.PackageManagement.NuGetPackageManager.<InstallPackageAsync>d__53.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at NuGet.CommandLine.InstallCommand.<InstallPackageAsync>d__38.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at NuGet.CommandLine.Command.Execute()
   at NuGet.CommandLine.Program.MainCore(String workingDirectory, String[] args)
```

nkolev92 commented:
Hm...is `\"` escaping the `"`? 
Extremely likely that's what happened. 

Can you try not including an trailing slash in the directory as you pass, it or not wrapping it in `"`. "

papeh commented:
You're right: the `\` was escaping the `\`. The command line parser is fancier than I expected.

We could leave the `.Trim('\"')` as a cushion for this case, or I'm happy to roll that back if you'd prefer clients be more rigorous.

```
C:\repos\repo\build>"C:\repos\repo\build\/nuget.exe" install BuildTasks -excludeVersion -version 2.3.1 -solutionDirectory "C:\repos\repo\build"
Feeds used:
  C:\Users\papeh\.nuget\packages\
  https://api.nuget.org/v3/index.json
  C:\Dev\nugetpackages
  C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\



Attempting to gather dependency information for package 'BuildTasks.2.3.1' with respect to project 'C:\repos\repo\build\packages', targeting 'Any,Version=v0.0'
Package 'BuildTasks 2.3.1' is not found in the following primary source(s): 'C:\Users\papeh\.nuget\packages\,https://api.nuget.org/v3/index.json,C:\Dev\nugetpackages,C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\'. Please verify all your online package sources are available (OR) package id, version are specified correctly.
```

nkolev92 commented:
Now that we know where the issue stems from, I think just improving the error message is probably better. 

Do you agree? cc @nuget/nuget-client

papeh committed:
Don't trim quotes from paths

papeh commented:
Agreed; I've rolled back quote trimming.

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx:
```diff
@@ -4433,6 +4439,12 @@ To prevent NuGet from downloading packages during build, open the Visual Studio
   <data name="FileConflictChoiceText_cht" xml:space="preserve">
     <value>[Y] 是  [A] 全部皆是  [N] 否  [L] 全部皆否 ?</value>
   </data>
+  <data name="Error_InvalidSolutionDirectory_deu" xml:space="preserve">
```
I think it's preferable if you don't add the localization yourself. 

Maybe @dominoFire can help get it reviewed by localization folks, but in general it's something something that may slow down reviews.

nkolev92 replied:
@papeh Can you please revert the localized versions of the strings? That way we can move forward toward merging this.

nkolev92 reviewed (commented):
Looks great to me, one question about the localization. Consider waiting on @dominoFire to respond.

jeffkl commented on src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs:
```diff
@@ -885,15 +885,25 @@ private void ProcessSolutionFile(string solutionFileFullPath, PackageRestoreInpu
-                restoreInputs.DirectoryOfSolutionFile,
-                NuGetConstants.NuGetSolutionSettingsFolder,
-                Constants.PackageReferenceFile);
+            try
```
I would only have this `try...catch` only around the `Path.Combine()` call, otherwise you might catch an `ArgumentException` that's not related (ie in `File.Exists()` or `Collection<T>.Add()`

nkolev92 replied:
@papeh Can you please address @jeffkl's feedback.

jeffkl reviewed (dismissed):
Thanks for the contribution, just one small comment to address from me.

papeh commented:
I'm out this week; hopefully back next week.

nkolev92 commented:
Sounds good, take your time @papeh 

We do appreciate the work you've done so far.

papeh committed:
Address code review comments

nkolev92 commented:
@papeh We had some CI issues, mind rebasing so I can merge?

papeh commented:
Sorry, I should have asked earlier if you wanted me to rebase.
@nkolev92, it's ready 😃

jeffkl commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs:
```diff
@@ -333,6 +333,21 @@ public void GetMsbuildDirectoryFromPath_PATHENVWithQuotes_Succeeds()
+        }
+
+        [Fact]
+        public void CombinePathWithVerboseError_IllegalCharacters_MessageContainsBadPath()
```
@papeh This unit test is failing on Mac, does it need to be updated to only run on Windows or can you make it fail on Mac?
```
2022-05-31T16:31:55.4018860Z [xUnit.net 00:00:43.41]     NuGet.CommandLine.Test.MsBuildUtilityTest.CombinePathWithVerboseError_IllegalCharacters_MessageContainsBadPath [FAIL]
2022-05-31T16:31:55.4753210Z   Failed NuGet.CommandLine.Test.MsBuildUtilityTest.CombinePathWithVerboseError_IllegalCharacters_MessageContainsBadPath [11 ms]
2022-05-31T16:31:55.4755930Z   Error Message:
2022-05-31T16:31:55.4757170Z    Assert.Throws() Failure
2022-05-31T16:31:55.4758490Z Expected: typeof(System.ArgumentException)
2022-05-31T16:31:55.4759940Z Actual:   (No exception was thrown)
2022-05-31T16:31:55.4761160Z   Stack Trace:
2022-05-31T16:31:55.4763360Z     at NuGet.CommandLine.Test.MsBuildUtilityTest.CombinePathWithVerboseError_IllegalCharacters_MessageContainsBadPath () [0x00009] in <42272c7ec5f748dc8bf025c2c1caf448>:0 
2022-05-31T16:31:55.4767360Z   at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)
2022-05-31T16:31:55.4769570Z   at System.Reflection.RuntimeMethodInfo.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x0006a] in <0e2013ab6b944e008349d8dcd759ab4d>:0 
```

papeh replied:
Thanks for that log. I've added a colon, which should throw on all platforms.

papeh committed:
Fix Illegal Path Character Unit Test

Add ':', the only illegal character on Mac OS

papeh commented:
@jeffkl, this builds ok for me (although I never got tests running, even before making changes). Maybe we could retry the build?

jeffkl commented:
> @jeffkl, this builds ok for me (although I never got tests running, even before making changes). Maybe we could retry the build?

Yes thanks for letting me know, I didn't get any notification that the build failed and forgot to check yesterday.  I've started a new build!

papeh commented:
And it failed on Mac again 😞. If the colon in the path didn't cause Path.Combine to throw, then your suggestion of skipping that test on Mac makes sense. I don't know how to implement that.

jeffkl commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MSBuildUtilityTest.cs:
```diff
@@ -333,6 +333,21 @@ public void GetMsbuildDirectoryFromPath_PATHENVWithQuotes_Succeeds()
+            Assert.Equal(Path.Combine(paths), MsBuildUtility.CombinePathWithVerboseError(paths));
+        }
+
+        [Fact]
```
This should make the test run on Windows only:

```suggestion
        [PlatformFact(Platform.Windows)]
```

papeh replied:
Thanks, @jeffkl. Looks like they passed on Linux, too, so I'll add that to the list.

jeffkl replied:
I don't think the test was running on Linux and for some reason its still running and failing on Mac.  Let's try Windows only as that still validates the `try...catch` logic and the functionality isn't OS specific so that should be fine.

papeh committed:
Skip Testing Path.Combine Exceptions on Mac

Even with what should be an illegal path character on Mac OS, Path.Combine doesn't throw.

papeh committed:
Test only Windows

For some reason, Platform.Linux seems to be causing this test to run on Mac

jeffkl reviewed (approved):
Thanks for the contribution @papeh!
